core/services/gas_updater.go

```diff
@@ -0,0 +1,140 @@
+package services
+
+import (
+	"fmt"
+	"math/big"
+	"sort"
+
+	"chainlink/core/eth"
+	"chainlink/core/logger"
+	"chainlink/core/store"
+	"chainlink/core/store/models"
+	"chainlink/core/utils"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+var (
+	promGasUpdaterAllPercentiles = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "gas_updater_all_gas_percetiles",
+		Help: "Gas price at given percentile",
+	},
+		[]string{"percentile"},
+	)
+
+	promGasUpdaterSetGasPrice = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "gas_updater_set_gas_price",
+		Help: "Gas updater set gas price (in Wei)",
+	},
+		[]string{"percentile", "block_num"},
+	)
+)
+
+//go:generate mockery -name GasUpdater  -output ../internal/mocks/ -case=underscore
+
+// GasUpdater listens for new heads and updates the base gas price dynamically
+// based on the configured percentile of gas prices in that block
+type GasUpdater interface {
+	store.HeadTrackable
+	RollingBlockHistory() []eth.Block
+}
+
+type gasUpdater struct {
+	store                   *store.Store
+	rollingBlockHistory     []eth.Block
+	rollingBlockHistorySize int
+	blockDelay              int64
+	percentile              int
+}
+
+// NewGasUpdater returns a new gas updater.
+func NewGasUpdater(store *store.Store) GasUpdater {
+	gu := &gasUpdater{
+		store:                   store,
+		rollingBlockHistory:     make([]eth.Block, 0),
+		rollingBlockHistorySize: int(store.Config.GasUpdaterBlockHistorySize()),
+		blockDelay:              int64(store.Config.GasUpdaterBlockDelay()),
+		percentile:              int(store.Config.GasUpdaterTransactionPercentile()),
+	}
+	return gu
+}
+
+func (gu *gasUpdater) Connect(bn *models.Head) error {
+	if gu.store.Config.GasUpdaterEnabled() {
+		logger.Debugw("GasUpdater: dynamic gas updates are enabled", "ethGasPriceDefault", gu.store.Config.EthGasPriceDefault())
+	} else {
+		logger.Debugw("GasUpdater: dynamic gas updating is disabled", "ethGasPriceDefault", gu.store.Config.EthGasPriceDefault())
+	}
+	return nil
+}
+
+func (gu *gasUpdater) Disconnect() {
+	return
+}
+
+// OnNewHead recalculates and sets global gas price on every head
+func (gu *gasUpdater) OnNewHead(head *models.Head) {
+	if !gu.store.Config.GasUpdaterEnabled() {
```

Comment:
It's safest to check this here, but it means we lose out on prometheus gas price reporting if it's disabled.

Reply:
It's safest to bail out as early as possible if the gas updater is disabled (so we avoid running any potentially destructive code).

The alternative would be to wrap only the actual call that updates the gas and leave the rest.

Reply:
I think that's fine.